File: 989.Add-to-Array-Form-of-Integer.go
// https://leetcode.com/problems/add-to-array-form-of-integer/
//
// algorithms
// Medium (43.39%)
// Total Accepted:    3,113
// Total Submissions: 7,175
// beats 100.0% of golang submissions

package leetcode

import "strconv"

func addToArrayForm(A []int, K int) []int {
	kArr := []byte(strconv.Itoa(K))
	aLen, kLen := len(A), len(kArr)

	carry := 0
	var res []int

	for i := 0; i < max(aLen, kLen); i++ {
		sum := carry
		if i < aLen {
			sum += A[aLen-1-i]
		}
		if i < kLen {
			sum += int(kArr[kLen-1-i] - '0')
		}
		carry = sum / 10
		sum %= 10
		res = append([]int{sum}, res...)
	}

	if carry > 0 {
		res = append([]int{carry}, res...)
	}

	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 989.Add-to-Array-Form-of-Integer.java
// https://leetcode.com/problems/add-to-array-form-of-integer/
//
// algorithms
// Medium (51.97%)
// Total Accepted: 27,452
// Total Submissions: 62,341
// beats 100.0% of java submissions


class Solution {
    public List<Integer> addToArrayForm(int[] A, int K) {
        int len = A.length;
        int idx = len - 1;
        int carry = 0;
        LinkedList<Integer> l = new LinkedList<>();

        while (K != 0 || idx >= 0) {
            int tmp = K % 10;
            K /= 10;
            tmp += carry;

            if (idx >= 0) {
                tmp += A[idx];
            }

            if (tmp >= 10) {
                tmp -= 10;
                carry = 1;
            } else {
                carry = 0;
            }
            l.offerFirst(tmp);

            idx--;
        }

        if (carry != 0) {
            l.offerFirst(carry);
        }

        return (List<Integer>) l;
    }
}